temporal-sdk/src/main/java/io/temporal/internal/sync/WorkflowRetryerInternal.java

```diff
@@ -197,31 +191,42 @@ private static RetryOptions getRetryOptionsSideEffect(String retryId, RetryOptio
             options.getMaximumAttempts(),
             maximumIntervalMillis,
             options.getDoNotRetry());
+
+    // Records retry options in the history allowing changing them without breaking determinism.
+    // We generate a callId deterministically once during the initial call to retryAsync and persist
+    // the original retryOptions in the side effect, so it can be looked up during replaying
     SerializableRetryOptions sRetryOptions =
         WorkflowInternal.mutableSideEffect(
-            retryId,
+            retryCallId,
             SerializableRetryOptions.class,
             SerializableRetryOptions.class,
-            Object::equals,
+            // we never want to persist a new value if there is already a previously saved value
```

Comment:
Why? I think it is fine to update the retry options during retries here. 

If we decide to never update retry options then `sideEffect` can be used instead of `mutableSideEffect`.

Reply:
Because it's how it works now.

You have SerializableRetryOptions without implemented equals (always returns false effectively in our usage)
Then you pass Object:equals as a function that defines when it should be updated (which is kinda crazy already, at least it should be "not equals"). So, you get false for always for it.
So, currently, we always pass false here effectively. And we always save the original retry options. I just expressed what we already have in an explicit manner and not bug-prone.